### DISCUSSION TITLE
Importing users and totp credentials?

### DISCUSSION THREAD

**Discussion by ryan-morris** [keycloak/keycloak#15334] [Q&A] [answered]
How is keycloak storing the OTP seed in the database?

I am migrating users over from ASPNET Core Identity. I have the original manual code / key / seed saved. (the same value as shown in the first image below). If I import that value I am unable to login with the authenticator app.  Looking at what is actually stored by keycloak, I'm just not sure what conversion I need to do.

For example, this code:

![ORIGINAL-CODE-SEED](https://user-images.githubusercontent.com/7102450/199777993-c5ef9907-4419-49f4-9e85-218199be8187.png)

Becomes this in the database:

![saved](https://user-images.githubusercontent.com/7102450/199777965-c1e97fe9-ab84-449a-a79b-dca932c80f6f.png)

What steps do I need to take to go from one to the other so that I can successfully import all the users?

Thanks

**Comment by darius-m** [comment]
I can't give you an exact code for it, but basically the value that is imported in the authenticator app is a base 32 encoding of the secret value. The QR code also contains some other information like username and SSO name that you need to manually set when using the base32-encoded string that you mention. See this [Google authenticator wiki page](https://github.com/google/google-authenticator/wiki/Key-Uri-Format) for more details.

**Reply by ryan-morris** [reply]
Keycloak:

- Builds the secret as a string based on alphanumeric values https://github.com/keycloak/keycloak/blob/main/server-spi/src/main/java/org/keycloak/models/utils/HmacOTP.java#L48
- That string is what is sent to the database
- The string / secret is Base32 encoded for display on the manual entry method 


My problem appears to be that other frameworks do not limit the secret to alphanumeric ascii characters. If I Base32 decode the seed from aspnet identity, then I get invalid ASCII characters which prevents Keycloak from letting me login.

Example::

```
var base32Str = "NA3GW6RYOB3GG2ZVOZFE4NSVM5ZEUVDS"; // base32 seed generated by keycloak during totp setup
var base32Bytes = Base32Encoding.ToBytes(base32Str); // base32 bytes
var seedAscii = System.Text.Encoding.ASCII.GetString(base32Bytes); // h6kz8pvck5vJN6UgrJTr 
```

```
var base32Str = "BHPPUOGKBU7O3OJHEPCJIWTX"; // base32 seed generated + stored by aspnet identity during setup
var base32Bytes = Base32Encoding.ToBytes(base32Str);
var seed = System.Text.Encoding.ASCII.GetString(base32Bytes); // 	??8?
                                                              // >?? '#??Zw
```


The fact keycloak only supports ASCII safe characters for the otp seed isn't condusive to importing from other systems

**Reply by darius-m** [reply]
After a quick look at the code, it does not appear that the code should have any issues if the value contains any non-alphanumeric characters (although it may be safer to only use those), it just generates the secrets using a limited character set.

Since [the credential validation code](https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/models/utils/CredentialValidation.java#L33) uses `getBytes()` to convert the key value it retrieves, you may be able to add binary characters as `\u00XY` (for HOTP the value is converted in the `generateOTP` function). You can try to see if you can manually set a secret value that contains non-alphanumeric characters in the database and confirm that it accepts your OTP codes.

**Reply by ryan-morris** [reply]
I was super hopeful about the binary characters but it doesn't seem to be working. It may be complicated by the fact it's a JSON string. I may be overlooking something here though. 

I have tried the following, all meant to represent the BHPPUOGKBU7O3OJHEPCJIWTX BASE32 seed.

These keep a-zA-Z0-9 as is, and \u character everything else. Only difference being the number of escape chars

`\u0009\u00de\u00fa56\u00ca\u000d\u003e\u00ed\u00b9\u0027\u0023\u00c4\u0094\u005a119`
`\\u0009\\u00de\\u00fa56\\u00ca\\u000d\\u003e\\u00ed\\u00b9\\u0027\\u0023\\u00c4\\u0094\\u005a119`
`\\\\u0009\\\\u00de\\\\u00fa56\\\\u00ca\\\\u000d\\\\u003e\\\\u00ed\\\\u00b9\\\\u0027\\\\u0023\\\\u00c4\\\\u0094\\\\u005a119`

These \u every character for consistency.  Only difference being the number of escape chars

`\u0009\u00de\u00fa\u0038\u00ca\u000d\u003e\u00ed\u00b9\u0027\u0023\u00c4\u0094\u005a\u0077`
`\\u0009\\u00de\\u00fa\\u0038\\u00ca\\u000d\\u003e\\u00ed\\u00b9\\u0027\\u0023\\u00c4\\u0094\\u005a\\u0077`
`\\\\u0009\\\\u00de\\\\u00fa\\\\u0038\\\\u00ca\\\\u000d\\\\u003e\\\\u00ed\\\\u00b9\\\\u0027\\\\u0023\\\\u00c4\\\\u0094\\\\u005a\\\\u0077`

native dotnet serialization with for utf8
`\t\uFFFD\uFFFD8\uFFFD\r\u003E\uFFFD\uFFFD\u0027#\u0114Zw`
`\\t\\uFFFD\\uFFFD8\\uFFFD\\r\\u003E\\uFFFD\\uFFFD\\u0027#\\u0114Zw`
`\\\\u0009\\\\u00de\\\\u00fa\\\\u0038\\\\u00ca\\\\u000d\\\\u003e\\\\u00ed\\\\u00b9\\\\u0027\\\\u0023\\\\u00c4\\\\u0094\\\\u005a\\\\u0077`

native dotnet serialization for ascii

`\\t??8?\\r\\u003E??\\u0027#??Zw`
`\t??8?\r\u003E??\u0027#??Zw`

```
                  id                  | salt | type |               user_id                | created_date  | user_label |                                            secret_data                                            |                               credential_data                                | priority
--------------------------------------+------+------+--------------------------------------+---------------+------------+---------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------+----------
 47002c2d-c07a-4533-b784-dffa9ee220a0 |      | otp  | 13cb2b42-73fe-4a80-b2aa-dd069e7ad837 | 1646254326239 |            | { "value": "\u0009\u00de\u00fa56\u00ca\u000d\u003e\u00ed\u00b9\u0027\u0023\u00c4\u0094\u005a119"} | {"digits":6,"counter":0,"period":30,"algorithm":"HmacSHA1","subType":"totp"} |       20
```

**Reply by ryan-morris** [reply]
one question though, is keycloak caching any user credential information? If I am manually issuing an update in postgres to the credential row in the database, is it actually reloading that to compare to the totp code? I've been logging in, getting to "One-time code" screen, trying a code, doing a database update, entering code, doing database update, etc. 

I tried restarting several times in between changes initially without any luck either just to be sure

**Reply by darius-m** [reply]
The first option is wrong - if you have any printable characters, you must use the equivalent ASCII character. By running
```bash
echo BHPPUOGKBU7O3OJHEPCJIWTX | base32 -d | xxd -g1
```
we get the following output
```
00000000: 09 de fa 38 ca 0d 3e ed b9 27 23 c4 94 5a 77     ...8..>..'#..Zw
```

Some of the characters are indeed printable - like `8`, `>` and so on. However, in your first representatation you have `\u0009\u00de\u00fa56...`, which are equivalent to `09 de fa 35 36`, since non-escaped characters are each treated as a full character in itself. The native dotnet serialization seems a bit odd - does it output some characters as negative integer values for some reason? (e.g. `\uFFFD`).

Regardless, you could try to test the conversion using some Java code to see if you get the expected binary values - e.g.
```java
String test = "\u0009\u00de\u00fa56`;
byte[] btest = test.getBytes();
for (int i = 0; i < btest.length; i++)
    System.out.printf("%02x ", btest[i]);
System.out.println("");
```

As an unfortunate note, the part of the code that performs serialization / deserialization from the database may escape the backslashes used for the unicode characters (e.g. `\u0009` would result the byte sequence `5c 75 30 30 30 39`), in which case I'm not sure if you can go for this approach. However, how the deserialization is handled is dependent on the library used, so you will need to test this in order to confirm.

**Reply by darius-m** [reply]
I am honestly not sure how credentials are cached. As far as I know, most systems do not cache this kind of data since it may cause issues with database synchronizations in clustered configurations, and it would not provide a great speed improvement since the cache would still be a database (functionally).

You can pretty easily test for caching by setting a known value in the database, logging in to confirm that it works, and then changing the value to anything else to confirm that it no longer works.

**Reply by ryan-morris** [reply]
Unfortunately, I have not had any luck. I feel like I've tried a hundred different combinations, these are some of what I have tried settings as the value in `{ "value": "" }` for the secret_data on the `credential` table.

I do need to restart keycloak in between changes by the way, which is kind of a pain in between attempts.

```c#
// BHPPUOGKBU7O3OJHEPCJIWTX = \u0009\u00de\u00fa8\u00ca\u000d>\u00ed\u00b9'#\u00c4\u0094Zw
public static string SeedToKeycloakMethod1(byte[] base32Bytes)
{
    StringBuilder sb = new StringBuilder();
    foreach (char c in base32Bytes)
    {
        var isOk = (int)c >= 32 && (int)c < 127;

        if (isOk)
        {
            sb.Append(c);
        }
        else
        {
            var escaped = String.Format(@"\u{0:x4}", (byte)c);
            sb.Append(escaped);
        }
    }

    var result = sb.ToString();
    return result;
}

// BHPPUOGKBU7O3OJHEPCJIWTX = \\u0009\\u00de\\u00fa8\\u00ca\\u000d>\\u00ed\\u00b9'#\\u00c4\\u0094Zw
public static string SeedToKeycloakMethod1a(byte[] base32Bytes)
{
    StringBuilder sb = new StringBuilder();
    foreach (char c in base32Bytes)
    {
        var isOk = (int)c >= 32 && (int)c < 127;

        if (isOk)
        {
            sb.Append(c);
        }
        else
        {
            var escaped = String.Format(@"\\u{0:x4}", (byte)c);
            sb.Append(escaped);
        }
    }

    var result = sb.ToString();
    return result;
}

// BHPPUOGKBU7O3OJHEPCJIWTX = \\\\u0009\\\\u00de\\\\u00fa8\\\\u00ca\\\\u000d>\\\\u00ed\\\\u00b9'#\\\\u00c4\\\\u0094Zw
public static string SeedToKeycloakMethod1b(byte[] base32Bytes)
{
    StringBuilder sb = new StringBuilder();
    foreach (char c in base32Bytes)
    {
        var isOk = (int)c >= 32 && (int)c < 127;

        if (isOk)
        {
            sb.Append(c);
        }
        else
        {
            var escaped = String.Format(@"\\\\u{0:x4}", (byte)c);
            sb.Append(escaped);
        }
    }

    var result = sb.ToString();
    return result;
}

// BHPPUOGKBU7O3OJHEPCJIWTX = \tÞú8Ê\r>í¹'#Ä\u0094Zw
public static string SeedToKeycloakMethod2(byte[] base32Bytes)
{
    StringBuilder sb = new StringBuilder();
    foreach (char c in base32Bytes)
    {
        var isOk = (int)c >= 32 && (int)c < 127;

        if (isOk)
        {
            sb.Append(c);
        }
        else
        {
            var input = ((int)c).ToString("x4");
            int p = int.Parse(input, System.Globalization.NumberStyles.HexNumber);
            sb.Append((char)p);
        }
    }

    var result = sb.ToString();
    return result;
}

// BHPPUOGKBU7O3OJHEPCJIWTX =  㣺්➹쐣媔 
public static string SeedToKeycloakMethod3(byte[] base32Bytes)
{
    return System.Text.Encoding.Unicode.GetString(base32Bytes);
}

// BHPPUOGKBU7O3OJHEPCJIWTX =  \t  8 \r>  '#ĔZw 
public static string SeedToKeycloakMethod4(byte[] base32Bytes)
{
    return System.Text.Encoding.UTF8.GetString(base32Bytes);
}

// \t??8?\r>??'#??Zw
public static string SeedToKeycloakMethod5(byte[] base32Bytes)
{
    return System.Text.Encoding.ASCII.GetString(base32Bytes);
}
```

**Reply by darius-m** [reply]
In this case, I've run out of ideas, sorry. Hopefully someone more knowledgeable can provide some input.

**Reply by ryan-morris** [reply]
Thanks so much for your help and suggestions.

**Reply by angelomelonas** [reply, marked as answer]
For anyone looking at a solution to this problem, see [this thread](https://github.com/keycloak/keycloak/pull/21117). When migrating credentials, you need to create a new Keycloak endpoint (or have another method of using the Keycloak Admin Client) and then create the credential for a user as follows:

```Java
OTPCredentialModel otpCredentialModel = OTPCredentialModel.createTOTP(
        otpSecret, // Migrated TOTP Secret
        6, // Digits
        30, // Period
        "HmacSHA1", // Algorithm
        OTPCredentialModel.SecretEncoding.BASE32.name() // Encoding. This is the important part.
);
new OTPCredentialProvider(session).createCredential(realmModel, userModel, otpCredentialModel);
```
